### PR TITLE
Openapi3 support for discriminated unions

### PR DESCRIPTION
**Pull request by mikekistler**
This PR adds support for the OpenAPI `discriminator` as a means to describe the type of polymorphism commonly referred to as "discriminated unions".

### CODE REVIEW

File: packages/openapi3/src/openapi.ts
Comment (markcowl) [inline packages/openapi3/src/openapi.ts:1181]: Do we want to allow string models and not string literals?  In this case, how would we enforce uniqueness among the discriminator values among the children?

File: packages/openapi3/test/test-openapi-output.ts
Comment (markcowl) [inline packages/openapi3/test/test-openapi-output.ts:387]: this is exactly the kind of discriminator I don't understand.  How on the wire can I discriminate between types?  Seems like we should require kind: "Cat"  and kind: "Dog"

File: packages/openapi3/test/test-openapi-output.ts
Comment (markcowl) [inline packages/openapi3/test/test-openapi-output.ts:422]: same comment

File: packages/openapi3/test/test-openapi-output.ts
Comment (markcowl) [inline packages/openapi3/test/test-openapi-output.ts:452]: don't feel like we want this - for a usable api, we have to be able to count on the discriminator values being known and unique

File: packages/openapi3/test/test-openapi-output.ts
Comment (markcowl) [inline packages/openapi3/test/test-openapi-output.ts:496]: same comment

File: packages/openapi3/test/test-openapi-output.ts
Comment (markcowl) [inline packages/openapi3/test/test-openapi-output.ts:541]: I *think* this one should cause an error diagnostic, for the same reasons as above

File: packages/openapi3/test/test-openapi-output.ts
Comment (markcowl) [inline packages/openapi3/test/test-openapi-output.ts:601]: nice

**Review by markcowl — CHANGES_REQUESTED** [review]
Mainly the implementation looks good.  IMO, we should be requiring the specification of discriminator values in child types (or at the very least, throwing a warning diagnostic if they are not present).

Also, think the decorator should not go into an emitter - which means that we should have a defined type for discrimnator information that emitters can use to describe or implement discriminated serialization / deserialization.

File: packages/openapi3/src/openapi.ts
Comment (mikekistler) [reply packages/openapi3/src/openapi.ts:73]: I don't share this view.  The "discriminator" is a uniquely OpenAPI construct.  Can you point to some other REST protocol that has the concept of "discriminator"?

File: packages/openapi3/src/openapi.ts
Comment (mikekistler) [reply packages/openapi3/src/openapi.ts:85]: Which emitters do you have in mind here?  Does gRPC have the concept of "discriminators"?

Comment (bterlson) [reply packages/openapi3/src/openapi.ts:85]: IMO discriminator is a property of unions, and is general purpose. If I want to support rust and TS, the natural thing to do I think is add discriminator to the union and synthesize that field in union variants when targeting languages with discriminated unions and no ADTs.

File: packages/openapi3/test/test-openapi-output.ts
Comment (mikekistler) [reply packages/openapi3/test/test-openapi-output.ts:387]: As an _API style rule_ I agree.  But I don't believe that these should be encoded into the "generic" openapi3 emitter.

Comment (bterlson) [reply packages/openapi3/test/test-openapi-output.ts:387]: I think I agree with the confusion. Maybe an example of how a generic string being a discriminator gets used in practice would help? What do existing openapi code generators give you on e.g. typescript for this pattern?

**Comment by mikekistler** [comment]
@markcowl Thank you for your comments!  But I think you have injected your notions of "good API design" (most of which I agree with) into the "generic" openapi3 emitter.  I think we should avoid imposing our views on API style at this level -- leave that to libraries built on top.

File: packages/openapi3/src/openapi.ts
Comment (mikekistler) [reply packages/openapi3/src/openapi.ts:85]: I don't mind moving this to the rest library.  I suppose it is generic enough in its current form.

File: packages/openapi3/test/test-openapi-output.ts
Comment (mikekistler) [reply packages/openapi3/test/test-openapi-output.ts:387]: A generic string is the "catch-all".  It lets me add lizard later if I want.

But I suppose restricting this now and then possibly loosening it up later is better than starting loose and then being unable to tighten later.  So I will make the change to require string constants and we'll see if we get feedback asking for generic string.

Comment (mikekistler) [reply packages/openapi3/test/test-openapi-output.ts:387]: An interesting twist here -- without subtype support, this preclude more than one level of inheritance.  E.g.:
```
      model Pet {
        name: string;
        weight?: float32;
      }                                                                                
      model Cat extends Pet {                                                                      
        kind: "cat";                                                                              
        meow: int32;                                                                               
      }                                                                                            
      model Dog extends Pet {                                                                      
        kind: "dog";                                                                              
        bark: string;                                                                              
      }
      model Beagle extends Dog {
        kind: "beagle";
        purebred: boolean;
      } 
```

I suppose that for now we can restrict this too, but I'm pretty sure folks will want multi-level inheritance.

**Comment by mikekistler** [comment]
Updates pushed.  Have a look and let me know if this addresses your concerns.

**Comment by timotheeguerin** [comment]
Can you with this change have nested discriminator if wanted? This is valid a valid pattern in OpenAPI if I rememeber correctly.

```cadl
@discriminator("kind")
model Pet {

}

model Cat  extends Pet {
  kind: "cat"
}

@discrimnator("breed")
model Dog extends Pet {
  kind: "dog"
}

model GermanShepard extends Dog {
   breed: "germanShepard"
}

```

File: packages/openapi3/test/test-openapi-output.ts
Comment (nguerrera) [reply packages/openapi3/test/test-openapi-output.ts]: I don't yet understand the connection between this and multiple levels. It feels like https://github.com/microsoft/cadl/pull/89#issuecomment-983775654 shows a pattern for multiple levels where the set of types is still closed?

**Comment by mikekistler** [comment]
Just re-pushed with updated logic to generate a diagnostic warning if a discriminated union variant does not define a discriminator property with a string literal value.

**Comment by mikekistler** [comment]
I investigated @timotheeguerin 's question on multi-level discriminators and that led to some adjustments in the support for multi-level inheritance.  I think both are now supported correctly.

File: packages/openapi3/src/openapi.ts
Comment (markcowl) [inline packages/openapi3/src/openapi.ts:821]: see my review on the cadl-azure pr.  I feel like this functionality ought to be moved into the extension defining the decorator, so it isn't duplicated across emitters.  I think both getting the discriminators, getting the child types, and getting the value -> type mapping shpould be common functions.    However, I don't think we need to hold up the PR for this.

File: packages/openapi3/src/openapi.ts
Comment (markcowl) [inline packages/openapi3/src/openapi.ts:922]: As before, I think some of this should be simplified by having more functionality in the extension implementing the decorator, although special validations will occur for many emitters, on top of basic validation (no system would allow a single key to map to multiple types, for example..

**Review by markcowl — CHANGES_REQUESTED** [review]
This looks great - some comments around placing parts of the implementation in the rest library, but this can be done in a later PR - none of these comments should block.

**Comment by mikekistler** [comment]
I just pushed a new (pre-squashed) commit that I think addresses all outstanding comments and adheres to the agreed design.  Please have a look and let me know if any further changes are needed.

**Review by markcowl — APPROVED** [review]
One thing: I think we should have a new swagger sample that uses discriminators as part of the samples converted to swagger on check-in, and documentation in the language walkthrough, but these can come in follow-up PRs.

**Comment by mikekistler** [comment]
I opened #151 and assigned myself to track the follow-on work.

**Pull request merged**
